node/src/components/block_proposer.rs

```diff
@@ -97,49 +98,120 @@ impl<REv> ReactorEventT for REv where
+        pending + proposed + finalized
+    }
+}
+
+mod prune {
+    use super::*;
+    /// Prunes expired deploy information from an individual DeployCollection, returns the total
+    /// deploys pruned
+    pub(super) fn prune_deploys(
+        deploys: &mut DeployCollection,
+        current_instant: Timestamp,
+    ) -> usize {
+        let initial_len = deploys.len();
+        deploys.retain(|_hash, header| !header.expired(current_instant));
+        initial_len - deploys.len()
+    }
+    /// Prunes expired deploy information from each ProtoBlockCollection, returns the total
+    /// deploys pruned
+    pub(super) fn prune_blocks(
+        blocks: &mut ProtoBlockCollection,
+        current_instant: Timestamp,
+    ) -> usize {
+        let mut pruned = 0;
+        let mut remove = Vec::new();
+        for (block_hash, deploys) in blocks.iter_mut() {
+            pruned += prune_deploys(deploys, current_instant);
+            if deploys.is_empty() {
+                remove.push(*block_hash);
+            }
+        }
+        blocks.retain(|k, _v| !remove.contains(&k));
+        pruned
+    }
+}
+
+/// Block proposer.
+#[derive(DataSize, Debug, Clone)]
+pub(crate) struct BlockProposer {
+    state: BlockProposerState,
+    #[data_size(skip)]
+    metrics: BlockProposerMetrics,
     // We don't need the whole Chainspec here (it's also unnecessarily big), just the deploy
     // config.
     #[data_size(skip)]
     chainspecs: HashMap<Version, DeployConfig>,
-    #[data_size(skip)]
-    metrics: DeployBufferMetrics,
 }
 
-impl DeployBuffer {
-    /// Creates a new, empty deploy buffer instance.
+impl BlockProposer {
+    /// Creates a new, block proposer instance with the provided internal state.
     pub(crate) fn new<REv>(
         registry: Registry,
         effect_builder: EffectBuilder<REv>,
-        finalized: ProtoBlockCollection,
+        state: BlockProposerState,
     ) -> Result<(Self, Effects<Event>), prometheus::Error>
     where
         REv: ReactorEventT,
     {
         let effects = effect_builder
-            .set_timeout(DEPLOY_BUFFER_PRUNE_INTERVAL)
+            .set_timeout(PRUNE_INTERVAL)
             .event(|_| Event::BufferPrune);
 
-        let pending = DeployCollection::default();
-        let proposed = ProtoBlockCollection::default();
-        let chainspecs: HashMap<Version, DeployConfig> = HashMap::new();
-        let metrics = DeployBufferMetrics::new(registry)?;
-        let this = DeployBuffer {
-            pending,
-            proposed,
-            finalized,
-            chainspecs,
+        let metrics = BlockProposerMetrics::new(registry)?;
+        let this = BlockProposer {
             metrics,
+            state,
+            chainspecs: HashMap::new(),
         };
         Ok((this, effects))
     }
 
-    /// Adds a deploy to the deploy buffer.
+    /// Get the serializable state from this `BlockProposer` instance.
+    pub(crate) fn get_state(&self) -> &BlockProposerState {
```

Comment:
```suggestion
    pub(crate) fn state(&self) -> &BlockProposerState {
```
(In Rust, `get_` is usually omitted from getters.)

Reply:
Sure, applied your suggestion.